scripts/validateDeploymentUtils.ts

```diff
@@ -0,0 +1,300 @@
+  const deploymentsPath = path.join(__dirname, "../deployments/" + hre.network.name);
+  const searchContractDeployment = checkAddressInFile(contractAddress);
+  const deployment = await searchDirectory(deploymentsPath, searchContractDeployment);
+  if (!deployment) {
+    throw new Error(`Could not find deployment ${contractAddress}`);
+  }
+  console.log("Deployment: " + deployment);
+  const contractName = path.basename(deployment, path.extname(deployment));
+  console.log("ContractName: " + contractName);
+  const deploymentJson = JSON.parse(fs.readFileSync(deployment, "utf-8"));
+  return {
+    contractName: contractName,
+    constructorArgs: deploymentJson.args,
+    deploymentTxHash: deploymentJson.transactionHash,
+  };
+}
+
+async function getArtifact(contractName: string): Promise<any> {
+  const findContract = findFile(contractName + ".json");
+  const buildPath = path.join(__dirname, "../artifacts/contracts/");
+  const searchResult = await searchDirectory(buildPath, findContract);
+  if (!searchResult) {
+    throw new Error("Artifact not found");
+  }
+
+  return JSON.parse(fs.readFileSync(searchResult, "utf-8"));
+}
+
+async function getBytecodeWithLinks(contractName: string): Promise<string> {
+  const artifact = await getArtifact(contractName);
+
+  // convert string to array for inplace replacement
+  const arrayBytecode = artifact.bytecode.split("");
+
+  for (const dependency of Object.values(artifact.linkReferences)) {
+    const dependencyName = Object.keys(dependency)[0];
+    const dependencyPositions = dependency[dependencyName];
+
+    const deploymentFilename = "./deployments/" + hre.network.name + "/" + dependencyName + ".json";
+    // extract deployed dependency address and trim 0x
+    const addr = (await getAddressFromDeployment(deploymentFilename)).substring(2);
+
+    for (const position of dependencyPositions) {
+      // calculate link index. Each byte is 2 symbols + 0x at the start.
+      const idx = 2 + position.start * 2;
+      arrayBytecode.splice(idx, 40, addr);
+    }
+  }
+
+  return arrayBytecode.join("");
+}
+
+async function compileContract(contractName: string) {
+  // Find artifact with our contract and remove it to force recompilation of this contract
+  const findContract = findFile(contractName + ".sol");
+  const buildPath = path.join(__dirname, "../artifacts/contracts/");
+  const searchResult = await searchDirectory(buildPath, findContract);
+  if (searchResult) {
+    fs.rmSync(searchResult, { recursive: true, force: true });
+  }
+
+  execSync("npx hardhat compile", { stdio: "inherit" });
+  console.log(`${contractName} compiled successfully.`);
+}
+
+//Using streaming read cause file can be big
+const getAddressFromDeployment = async (filename: string): Promise<string | null> => {
+  if (fs.lstatSync(filename).isDirectory()) {
+    return null;
+  }
+
+  try {
+    const fileStream = fs.createReadStream(filename);
```

Comment:
wondering would that work?

```
return JSON.parse(fs.readFileSync(filename)).address ?? null
```

or could you explain why do you need to process it this way?

Reply:
i think you can also skip all files that don't end with ".json" (with this checking isDirectory wont be necessary)